### ISSUE TITLE
Handshake failed with fatal error SSL_ERROR_SSL: error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER.

### ISSUE THREAD

**Issue by cavator** [grpc/grpc#32076]
guys i'll post here all the things to reproduce my scenario so someone can try to really help me
this is the commands i used to generate the certs, i was following this tuto from [RealPython](https://realpython.com/python-microservices-grpc/#server-authentication)

```bash
openssl req -x509 -nodes -newkey rsa:4096 -keyout ca.key -out ca.pem -subj /O=me

openssl req -nodes -newkey rsa:4096 -keyout server.key -out server.csr -subj /CN=recommendations

openssl x509 -req -in server.csr -CA ca.pem -CAkey ca.key -set_serial 1 -out server.pem
```

this is my python server
```python
     async def serve():
        conn: Connection = await db_connect()
        interceptors = [AsyncExceptionToStatusInterceptor()]
        server = grpc.aio.server(futures.ThreadPoolExecutor(max_workers=10), interceptors=interceptors)
        add_UserServicer_to_server(UserService(conn), server)
        srv_key = os.getenvb(b'SERVER-KEY')
        srv_cert = os.getenvb(b'SERVER-CERT')
        creds = grpc.ssl_server_credentials([(srv_key, srv_cert)])
        listen_addr = '[::]:443'
        server.add_secure_port(listen_addr, creds)
        logging.info('Starting server on %s', listen_addr)
        await server.start()
        await server.wait_for_termination()
        await conn.close()
```
the Dockerfile

```dockerfile
FROM gcr.io/google_appengine/python

RUN virtualenv -p python3.7 /env

ENV VIRTUAL_ENV /env
ENV PATH /env/bin:$PATH

RUN mkdir /user
COPY grpc/py/user/ /user/

WORKDIR /user
RUN pip install -r requirements.txt
RUN apt-get install -y libpq5

EXPOSE 443

ENTRYPOINT ["python", "-u", "./user.py"]
```
the requirements.txt

```
grpc-interceptor ~=0.15.0
grpcio-tools==1.51.1
grpcio-status==1.51.1
grpcio-reflection==1.51.1
googleapis-common-protos==1.58.0
pyjwt==2.6.0
asyncpg==0.27.0
requests ~= 2.28.1
cryptography==39.0.0
```

a minimal .proto file to the lazy ones kkkkk
```proto
syntax = "proto3";
package app.project;

service User {
  rpc GetUser(IdRequest) returns (UserResponse);
}
message IdRequest{
  string id = 1;
}
message UserResponse {
  string uid = 1;
  string display_name = 2;
  string email = 3;
}
```

the api_config.yaml

```yaml
type: google.api.Service
config_version: 3
name: "*.apigateway.project.cloud.goog"
title: User Service API
apis:
  - name: app.project.User
usage:
  rules:
  - selector: app.derole.User.GetUser
    allow_unregistered_calls: true
backend:
  rules:
    - selector: "*"
      address: grpcs://user-service-{IP}-uc.a.run.app
```
the gcloud commands i use to generate the API-CONFIG and API-GATEWAY to the Run service
```bash
python3 -m grpc_tools.protoc  --include_imports --include_source_info --proto_path=../proto --descriptor_set_out=../proto/api_descriptor.pb --python_out=../generated_pb2 --grpc_python_out=../generated_pb2 user.proto
gcloud api-gateway api-configs create grpc-user-config  --api=grpc-user     --project={YOUR PROJECT}    --grpc-files=../proto/api_descriptor.pb,../api_config/user_api_config.yaml
gcloud api-gateway gateways create grpc-useR --api=grpc-user --api-config=grpc-user-config --location=us-central1 --project={YOUR PROJECT}
```

now u can reproduce a minimal thing  what i'm doing, so if someone see what i'm missing i would be glad to fix it

**Comment by buzzsurfr** [comment]
Because the error involves openssl, I would try using the openssl CLI to establish the TLS connection in the same manner and see if you get any more detail out of the attempt. The command would look similar to:

```
openssl s_client -connect user-service-ip-uc.a.run.app:443 ...
```

One other avenue to check would be the use of the hyphen/dash in the environment variable name. Maybe switch to underscore or add code to validate the key and certificate match before starting the server.

**Comment by gnossen** [comment]
@ZhenLian What TLS version does the library expect?

**Comment by ZhenLian** [comment]
Sorry for the late reply, I am OOO right now. Adding @erm-g to take a look, thanks!

**Comment by mianowski** [comment]
I've got the same issue, when trying to connect to SpiceDB and still look for a good solution

**Comment by gnossen** [comment]
@erm-g Friendly ping.

**Comment by XuanWang-Amos** [comment]
Checked with Andrey, we'll pick this up in July, please comment here if this is impacting your use case too, we'll adjust the priority accordingly.

**Comment by sonishi85** [comment]
Using grpcio-tools 1.53.0, and even with no connections on the channel, I see these errors every 10s:

0608 12:53:34.109768247      97 ssl_transport_security.cc:1420]       Handshake failed with fatal error SSL_ERROR_SSL: error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER.
E0608 12:53:44.109623375      96 ssl_transport_security.cc:1420]       Handshake failed with fatal error SSL_ERROR_SSL: error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER.
E0608 12:53:54.111122107     106 ssl_transport_security.cc:1420]       Handshake failed with fatal error SSL_ERROR_SSL: error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER.
E0608 12:54:04.110720553      96 ssl_transport_security.cc:1420]       Handshake failed with fatal error SSL_ERROR_SSL: error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER.

I thought maybe this was tied to my gRPC KA configuration but I disabled that and the pattern still occurs.

What's puzzling is that when a client does connect to the server, it is able to connect w/TLS and perform some operations though the channel is not seem stable and the connection drops after a while.

**Comment by coffeegist** [comment]
i am also experiencing this

**Comment by matthewstevenson88** [comment]
The two common ways to get this error are (a) the client does not use SSL/TLS when attempting to connect to the server and the server requires SSL/TLS, and (b) the client attempts to connect to the server using a TLS version that the server does not support.

Currently, gRPC C++ and Python servers exclusively support TLS 1.2 and TLS 1.3.

In the original comment from @cavator, it sounds like there is a proxy between the end user who is running `grpcurl` and the Cloudrun app? If that's correct, is it possible that the proxy is not talking SSL/TLS to your Python gRPC server?

@sonishi85 Is it possible that there is something trying to probe your server (e.g. health checking or telemetry) every 10 seconds? And possibly the thing doing the probing is trying to talk plaintext HTTP or gRPC, rather than SSL/TLS?

@coffeegist Can you provide more information about your setup?

Thanks all for your help. I think the key to resolving this issue is going to be providing a minimal reproducible example that is outside of any of your production environments, because this kind of issue is so dependent on the client behavior.

**Comment by Amoln123** [comment]
E0000 00:00:1730802713.715037   12832 ssl_transport_security.cc:1654] Handshake failed with fatal error SSL_ERROR_SSL: error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED

Please help me how to resolve this

**Comment by matthewstevenson88** [comment]
@Amoln123 The `CERTIFICATE_VERIFY_FAILED` error indicates that your gRPC application failed to successfully verify the peer's certificate, which is different than the `WRONG_VERSION_NUMBER` error under discussion here. If your peer certificates can be verified successfully by other stacks (e.g. the openssl command line) but not by gRPC, please open a new issue and we can help you debug.

**Comment by chrbsg** [comment]
Just a possible hint for anyone who lands on this page. I had this error - both `handshake read failed` and `OPENSSL_internal:WRONG_VERSION_NUMBER` . Also the connection was unstable in the sense that sometimes the client worked, and sometimes not. The problem turned out to be that multiple grpc server processes were running at the same time on the *same* port. As grpc uses the SO_REUSEPORT socket option by default, if multiple server processes are started, they will bind to the same port and service alternating requests (with the kernel as a primitive load balancer), hence the unstable client behaviour. The `handshake read failed` I suspect was caused by the client connecting to a server process that had blocked while shutting down.

**Comment by Anveena** [comment]
> E0000 00:00:1730802713.715037 12832 ssl_transport_security.cc:1654] Handshake failed with fatal error SSL_ERROR_SSL: error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
> 
> Please help me how to resolve this

Cool dude, I ran into the exact same problem. Did you solve it? How did you fix it?

**Comment by ImadSaddik** [comment]
In my case, I was getting this error when trying to use `ChatVertexAI` from `langchain_google_vertexai`

```text
I0000 00:00:1750671744.187049 10983 ssl_transport_security.cc:1665] Handshake failed with error SSL_ERROR_SSL: error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED: unable to get local issuer certificate
```

I was able to solve the issue by updating the CA certificates

```bash
sudo apt-get update
sudo apt-get install ca-certificates
sudo update-ca-certificates
```

And upgrading `certifi`

```bash
pip install --upgrade certifi
```
